graphbolt/src/fused_csc_sampling_graph.cc

```diff
@@ -484,6 +495,299 @@ auto GetTemporalPickFn(
+                          (hetero_with_seed_offsets)
+                              ? std::upper_bound(
+                                    seed_offsets->begin(), seed_offsets->end(),
+                                    i) -
+                                    seed_offsets->begin() - 1
+                              : 0;
+                      // `seed_index` indicates the index of the current
+                      // seed within the group of seeds which have the same
+                      // node type.
+                      const auto seed_index =
+                          (hetero_with_seed_offsets)
+                              ? i - seed_offsets->at(seed_type_id)
+                              : i;
+                      num_pick_fn(
+                          offset, num_neighbors,
+                          num_picked_neighbors_data_ptr + 1, seed_index,
+                          etype_id_to_num_picked_offset);
+                    }
+                  });
+
+              if (hetero_with_seed_offsets) {
+                torch::Tensor num_picked_offset_tensor =
+                    torch::zeros({num_etypes + 1}, indptr_options);
+                torch::Tensor substract_offset =
+                    torch::zeros({num_etypes}, indptr_options);
+                const auto substract_offset_data_ptr =
+                    substract_offset.data_ptr<indptr_t>();
+                const auto num_picked_offset_data_ptr =
+                    num_picked_offset_tensor.data_ptr<indptr_t>();
+                for (auto i = 0; i < num_etypes; ++i) {
+                  num_picked_offset_data_ptr[i + 1] =
+                      etype_id_to_num_picked_offset[i + 1];
+                  // Collect the total pick number for each edge type.
+                  if (i + 1 < num_etypes)
+                    substract_offset_data_ptr[i + 1] =
+                        num_picked_neighbors_data_ptr
+                            [etype_id_to_num_picked_offset[i]];
+                  num_picked_neighbors_data_ptr
+                      [etype_id_to_num_picked_offset[i]] = 0;
+                }
+                substract_offset =
+                    substract_offset.cumsum(0, indptr_.scalar_type());
+                subgraph_indptr_substract = ops::ExpandIndptr(
+                    num_picked_offset_tensor, indptr_.scalar_type(),
+                    substract_offset);
+              }
+
+              // Step 2. Calculate prefix sum to get total length and offsets of
+              // each node. It's also the indptr of the generated subgraph.
+              subgraph_indptr = num_picked_neighbors_per_node.cumsum(
+                  0, indptr_.scalar_type());
+              auto subgraph_indptr_data_ptr =
+                  subgraph_indptr.data_ptr<indptr_t>();
+
+              // When doing non-temporal hetero sampling, we generate an
+              // edge_offsets tensor.
+              if (type_per_edge_.has_value() && hetero_with_seed_offsets) {
```

Comment:
Do we need type_per_edge_ check here? If hetero_with_seed_offsets is true, then it should already be a valid tensor.